forge-ai/src/main/java/forge/ai/ComputerUtilMana.java

```diff
@@ -1689,12 +1689,16 @@ private static ListMultimap<Integer, SpellAbility> groupSourcesByManaColor(final
      * @since 1.0.15
      */
     public static int determineLeftoverMana(final SpellAbility sa, final Player player, final boolean effect) {
-        for (int i = 1; i < 100; i++) {
+        int max = 99;
+        if (sa.hasParam("XMaxLimit")) {
+            max = Math.min(max, AbilityUtils.calculateAmount(sa.getHostCard(), sa.getParam("XMaxLimit"), sa));
+        }
+        for (int i = 1; i <= max; i++) {
```

Comment:
Should we do a binary search instead of a linear search?

Reply:
what is your suggestion?